feat: Security Policy for Listener in Application load balancer

KollaAdithya opened the pull request:
<!-- Provide summary of changes -->
This PR resolves #1342
AWS Load Balancer listeners can restrict the tls protocol versions clients use to connect through an [sslpolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-listener.html#cfn-elasticloadbalancingv2-listener-sslpolicy). The default policy allows for insecure protocol versions for compatibility.
For those with stricter security goals it would be great to expose the ability to only accept current tls protocol versions.

This security policy for a HTTPS listener helps to negotiate SSL connections between a client and the load balancer.

For Example if Environment manifest contains,
#### In case of Public Application load balancer
```yaml
http:
  public: 
    certificates:
      - arn:aws:acm: ${AWS_REGION}:${AWS_ACCOUNT_ID}:certificate/13245665-cv8f-adf3-j7gd-adf876af95
    ssl_policy: ELBSecurityPolicy-FS-1-1-2019-08
```
#### In case of Internal load balancer
```yaml
http:
 private: 
    certificates:
      - arn:aws:acm: ${AWS_REGION}:${AWS_ACCOUNT_ID}:certificate/13245665-cv8f-adf3-j7gd-adf876af95
    ssl_policy: ELBSecurityPolicy-FS-1-1-2019-08
```
By Default `ELBSecurityPolicy-2016-08` will be set for a https listener in a public application load balancer and internal load balancer. You can access the available security policies at [https://docs.aws.amazon.com/elasticloadbalancing/latest/application/create-https-listener.html#describe-ssl-policies](url)


<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

KollaAdithya committed:
Security Policy for Listener

paragbhingre commented on internal/pkg/manifest/env_test.go:
```diff
@@ -797,11 +797,16 @@ func TestPublicHTTPConfig_IsEmpty(t *testing.T) {
 			wanted: true,
 		},
-		"not empty": {
+		"not empty when Certificates are attached ": {
```
```suggestion
		"not empty when Certificates are attached": {
```

paragbhingre commented on internal/pkg/deploy/cloudformation/stack/env.go:
```diff
@@ -405,13 +405,15 @@ func (e *EnvStackConfig) publicHTTPConfig() (template.HTTPConfig, error) {
 	return template.HTTPConfig{
 		ImportedCertARNs: e.importPrivateCertARNs(),
 		CustomALBSubnets: e.internalALBSubnets(),
+		SecurityPolicy:   e.in.Mft.EnvironmentConfig.HTTPConfig.Private.SecurityPolicy,
```
We can also do `e.getSecurityPolicy()` returning the value of `e.in.Mft.EnvironmentConfig.HTTPConfig.Private.SecurityPolicy` so that it is consistent with rest 2.

paragbhingre commented on internal/pkg/deploy/cloudformation/stack/testdata/environments/template-with-imported-certs-observability.yml:
```diff
@@ -521,6 +521,7 @@ Resources:
       LoadBalancerArn: !Ref PublicLoadBalancer
       Port: 443
       Protocol: HTTPS
+      SslPolicy: 'ELBSecurityPolicy-2016-08'
```
Can we also have at lease one test case that uses custom value provided by the customer?

paragbhingre reviewed (commented):
Overall looks great, just few nits.

KollaAdithya committed:
address fb

paragbhingre committed:
fix: Env manifest UX fix for http config  (#4068)

This PR fix multiple problems -
1. This PR fix the env manifest fields that were organized in incorrect manner.
2. Also resolves #4032
3. `copilot env init` with private subnets create env manifest with following field
```
http:
  private:
```
This should only be created when there is at least one parameter inside private is enabled. This PR fix this nit as well.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/env_integration_test.go:
```diff
@@ -292,6 +292,38 @@ network:
 			}(),
 			wantedFileName: "template-with-importedvpc-flowlogs.yml",
 		},
+		"generate template with public application load balancer with out a default security policy": {
```
Can we combine this test case into some existing ones (if they are independent with each other)? Otherwise if anything in our template changes, we will end up updating tens of test files...

KollaAdithya replied:
I am combining this test case with one of the existing ones

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/env_integration_test.go:
```diff
@@ -292,6 +292,38 @@ network:
 			}(),
 			wantedFileName: "template-with-importedvpc-flowlogs.yml",
 		},
+		"generate template with public application load balancer with out a default security policy": {
```
```suggestion
		"generate template with public application load balancer without a default security policy": {
```

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/env_integration_test.go:
```diff
@@ -292,6 +292,38 @@ network:
+  public:
+    certificates:
+      - cert-1
+    security_policy: ELBSecurityPolicy-FS-1-1-2019-08`
```
I don't see `ELBSecurityPolicy-FS-1-1-2019-08` in any of the updated expected CFN template file 🤔

KollaAdithya replied:
With the updated changes you can find `ELBSecurityPolicy-FS-1-1-2019-08` at `internal/pkg/deploy/cloudformation/stack/testdata/environments/template-with-imported-certs-sslpolicy-custom-empty-security-group.yml`

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/testdata/environments/template-with-imported-certs-observability.yml:
```diff
@@ -577,6 +578,7 @@ Resources:
       LoadBalancerArn: !Ref InternalLoadBalancer
       Port: 443
       Protocol: HTTPS
+      SslPolicy: 'ELBSecurityPolicy-2016-08'
```
Why are we changing the CFN template when the corresponding manifest is not updated?

iamhopaul123 commented on internal/pkg/template/templates/environment/cf.yml:
```diff
@@ -432,6 +433,7 @@ Resources:
       LoadBalancerArn: !Ref InternalLoadBalancer
       Port: 443
       Protocol: HTTPS
+      SslPolicy: {{ if .PrivateHTTPConfig.SecurityPolicy }}{{ .PrivateHTTPConfig.SecurityPolicy }}{{ else }} ELBSecurityPolicy-2016-08 {{ end }} 
```
Can we make the default behavior the same? aka if the field is not set we just don't set `SslPolicy` at all.

paragbhingre replied:
+1

iamhopaul123 commented on internal/pkg/template/env.go:
```diff
@@ -135,6 +135,7 @@ type HTTPConfig struct {
 	ImportedCertARNs  []string
 	CustomALBSubnets  []string
 	ELBAccessLogs     *ELBAccessLogs
+	SecurityPolicy    *string
```
It seems like we never assign value for this field. Ideally we should pass manifest value to this field otherwise it will always stay `nil`.

KollaAdithya replied:
In `internal/pkg/deploy/cloudformation/stack/env.go` we are passing manifest value to this field with the help of two functions that returns manifest value.
`func (e *EnvStackConfig) getPublicSslPolicy() *string `
`func (e *EnvStackConfig) getPrivateSslPolicy() *string`

huanjani commented on internal/pkg/manifest/env.go:
```diff
@@ -434,6 +434,7 @@ type PublicHTTPConfig struct {
 	SecurityGroupConfig ALBSecurityGroupsConfig `yaml:"security_groups,omitempty"`
 	Certificates        []string                `yaml:"certificates,omitempty"`
 	ELBAccessLogs       ELBAccessLogsArgsOrBool `yaml:"access_logs,omitempty"`
+	SecurityPolicy      *string                 `yaml:"security_policy,omitempty"`
```
Can we make this consistent with https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/manifest/lb_web_svc.go#L231 so that both manifests have the field named [`ssl_policy`](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#nlb-ssl-policy)`?

KollaAdithya replied:
yes, I can do that

huanjani commented on internal/pkg/deploy/cloudformation/stack/testdata/environments/template-with-basic-manifest.yml:
```diff
@@ -725,6 +725,7 @@ Resources:
       LoadBalancerArn: !Ref PublicLoadBalancer
       Port: 443
       Protocol: HTTPS
+      SslPolicy: 'ELBSecurityPolicy-2016-08'
```
Because you have added the template with ALB security policy, we probably don't need to add it to all of these other templates, especially the basic manifest one.

huanjani replied:
Oh, nvm. I see-- this is the default. Because it's not a required field, can we just not include it if the user doesn't specify a non-default policy?

KollaAdithya replied:
sure I will change accordingly

KollaAdithya committed:
address fb iamhopaul123 and huanjani

huanjani committed:
chore(pipelines): remove here string from buildspec (#4102)

Fixes second issue in #3553.

Here strings (`<<<`) don't work with `sh`. This change makes the pipeline buildspec a bit more universal.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

huanjani commented on internal/pkg/deploy/cloudformation/stack/env.go:
```diff
@@ -405,13 +405,15 @@ func (e *EnvStackConfig) publicHTTPConfig() (template.HTTPConfig, error) {
 		CIDRPrefixListIDs: e.in.CIDRPrefixListIDs,
 		ImportedCertARNs:  e.importPublicCertARNs(),
 		ELBAccessLogs:     elbAccessLogsConfig,
+		SSLPolicy:         e.getPublicSslPolicy(),
```
nit (here and elsewhere):
```suggestion
		SSLPolicy:         e.getPublicSSLPolicy(),
```
because https://github.com/golang/go/wiki/CodeReviewComments#initialisms

KollaAdithya committed:
fix merge conflicts

KollaAdithya committed:
Merge branch 'mainline' into sslpolicy

KollaAdithya committed:
address feedback

huanjani commented on internal/pkg/template/templates/environment/cf.yml:
```diff
@@ -356,6 +356,9 @@ Resources:
       Port: 443
       Protocol: HTTPS
+{{ if .PublicHTTPConfig.SSLPolicy }}
+      SslPolicy: {{ .PublicHTTPConfig.SSLPolicy }}
```
nit (here and below, and above in integ tests):
```suggestion
      SSLPolicy: {{ .PublicHTTPConfig.SSLPolicy }}
```

huanjani commented on internal/pkg/deploy/cloudformation/stack/env.go:
```diff
@@ -524,3 +526,11 @@ func (e *EnvStackConfig) internalALBSubnets() []string {
+}
+
+func (e *EnvStackConfig) getPrivateSSLPolicy() *string {
+	return e.in.Mft.EnvironmentConfig.HTTPConfig.Public.SSLPolicy
```
??
```suggestion
	return e.in.Mft.EnvironmentConfig.HTTPConfig.Private.SSLPolicy
```

huanjani commented on internal/pkg/template/templates/environment/cf.yml:
```diff
@@ -432,6 +435,9 @@ Resources:
       LoadBalancerArn: !Ref InternalLoadBalancer
       Port: 443
       Protocol: HTTPS
+{{ if .PublicHTTPConfig.SSLPolicy }}
+      SslPolicy: {{ .PublicHTTPConfig.SSLPolicy }}
```
```suggestion
{{ if .PrivateHTTPConfig.SSLPolicy }}
      SSLPolicy: {{ .PrivateHTTPConfig.SSLPolicy }}
```

huanjani reviewed (approved):
Looks good! 😻 Just a few changes, please, then feel free to remove the do-not-merge label! Thank you!

KollaAdithya committed:
address fb @janice
